### ISSUE TITLE
ui-template support for "Reload last value on refresh"

### ISSUE THREAD

**Issue by i8beef** [FlowFuse/node-red-dashboard#428]
### Description

Saw the `script` support was added, thank you!

The old dashboard templates had an options, "Reload last value on refresh". I believe this replayed the last `msg`, as all JS events defined in the template would get refired. This is useful to load an initial state for widgets. I am not sure how you are storing state like this across the nodes, but imagine all nodes actually would benefit from this generically to set initial states... if so, widen to include ALL controls? 😄 

### Have you provided an initial effort estimate for this issue?

I am no FlowFuse team member

**Comment by omrid01** [comment]
Replay of previous messages is tricky. For example, some messages could serve as action triggers, which should not be replayed.
There are many issues with this in Dashboard 1.0, including replay of messages before the UI is fully initialized, and unsolicited replay of the last incoming message, even when the checkbox is unchecked.
If this option is to be introduced into Dashboard 2.0, it **must** allow to be turned off in the node configuration.

**Comment by i8beef** [comment]
Absolutely, looking for exactly the same behavior as dashboard 1 here where there's a checkbox at the bottom of the settings screen for this. Without this, you have no way for your template to reflect "current state" when you reload the page, etc. It looks like most of the other controls already sort of do this for the same reason, otherwise anytime you reload the page everything would reset.

I'll leave implementation details up to you guys here, as I'm not familiar with the underlying approaches you already take for that, but I would imagine it would be extending whatever approach you use for the other controls, to the ui-template control (and any others that I haven't explored that don't currently do this).

**Comment by i8beef** [comment]
Ok I can state this better now. `ui-remplate` restores the last `<template>` state as expected... where this gets tricky is that if you are building a more involved widget, its likely the `<script>` section will make DOM changes, or changes to the `data` of the widget, etc., and THIS state is NOT restored.

Rather than using my own abstract case as an example, I'm gonna change it to the "counter" example in the documentation as it demonstrates this. You can click that button several times, reload the page, and the counter resets. It would be nice if that state persisted.

I would be fine with having to work within that "custom widget" framework to persist a widget state. I was just using the `this.$socket.on('msg-input:' + this.id, function(msg) { ... })` plus `setInterval` trick to attach DOM elements (I use it to load up a `video` element after a third party library loads), thus why I was talking about msg replays: if "state restore" here meant "replay last msg" then you get `<template>` and `<script>` execution changes all in one go.

I have an alternative for my SPECIFIC issue already (I got lucky though with how my external lib makes use of custom HTML elements and when element onInit's get called being agnostic to when the element was DEFINED), but hopefully this lays out the case for why I was asking about state restore and custom widget state here.

Essentially, restore of JUST the last evaluated `<template>` seems to miss results of `<script>` execution / "custom widget" state changes. I dont know what the right design is here that would allow for all that, but thats the general PROBLEM I'm concerned with.

**Comment by omrid01** [comment]
I agree with the comments above. Restoring the state of dashboard objects could mean different things for different implementations. Whatever the implementation will be in dashboard 2.0, we must have the ability to **turn off** the automatic restoration, in case it is not suitable. This ability (to disable auto-restoration) should go beyond ui-template, to EVERY dashboard widget.
For example, in my v1.0 dashboards I have text fields showing the status of executed actions (success, fail etc.). When I load the page I get the status of the previous action, which is irrelevant (and I need to explicitly clean it up upon page load, which is annoying).

**Comment by colinl** [comment]
@i8beef, @omrid01 is there agreement on *exactly* what is required to fix this issue?

**Comment by i8beef** [comment]
Its been 3 years so Im just going off of past-i8beef's description here cause 3-year-older me's memory sucks.... I definitely am still using a trick in my video player widgets with `this.$socket.on('msg-input:' + this.id, function(msg) { })`, but I think its just to setup a timer for initializing the video players because a a supporting library script is loaded in by a separate ui-template and I needed to wait for that to be available before initializing them...

I think when I reloaded the page before on those, it would restore the `<template>` content, but since the actual `<script>` content wouldn't get RETRIGGERED it wouldn't load the video player, but Im unsure... today, it seems to be working though, if I reload that page the video player gets re-inited and comes up ok, so I obviously found a way to make it work, OR you guys already changed something that gave me this in another way.

**Comment by colinl** [comment]
The last message should get sent to the template on refresh, but this was always the case as far as I remember, so I am not sure exactly what you were asking for.

**Comment by omrid01** [comment]
> [@i8beef](https://github.com/i8beef), [@omrid01](https://github.com/omrid01) is there agreement on _exactly_ what is required to fix this issue?

There were long debate conversations about the widget restoration issue in the Node-red forum. Many people do not understand the potential issues it may cause. For example, restoring the last value of a text field could be beneficial, but irrelevant/annoying in a text input field. Let alone replaying a message which invokes a action (e.g. "pay now"), which will be re-invoked upon every screen reload.
Aa far as I remember, in dashboard-2 there are 2 restoration repositories, one for "last status" (last message) and one for the event sequence, but I don't know if & how they are applied automatically. 
In my own dashboard-2 table node, I decided not to depend on any "behind the scenes" auto-restoration. instead, I am implementing this by myself (reading from & writing to Node-red's state repository, and populating the data in the widget).

**Comment by colinl** [comment]
@omrid01 I am trying to understand whether there is anything specific that needs changing here in order to determine whether it will be fixed by the ongoing work in #2234.

**Comment by i8beef** [comment]
Ok... so I might be looking at my WORKAROUND for whatever the issue was then. In a template node `<script>` tag I have essentially this, which pulls some timer based shenanigans to get around the fact that I couldn't reliably hook the load of a script in ANOTHER node (so wait until its available to init):

```
<script>
    this.$socket.on('msg-input:' + this.id, function(msg) {
        let timer = setInterval(function() {
            if (!window.VideoStream || msg === undefined || msg.payload === undefined) return;
            clearInterval(timer);

            const container = document.getElementById("camera1WebRtcVideo");
            if (container.hasChildNodes()) {
                container.removeChild(container.lastElementChild);
            }

            const video = document.createElement("video-stream");
            video.background = false;
            video.mode = "webrtc" || video.mode;
            video.src = new URL(msg.payload);
            container.appendChild(video);
        }, 100);
    });
</script>
```

I can reload this page and it reinits the video players, so either putting a few things together, Im betting that my first attempt was reliant on that `script` section actually RUNNING on reload, which would require a msg republish... but that I must have changed this in a way that allowed the video player script to be solely HTML reliant instead of requiring a JS init, so just restoring the `template` contents (post-init of the  initial run) was enough to get me working.

As such I dont think I explicitly need this anymore for my specific case... but does my original ask still stand for other cases? If someone were making a custom widget in a ui-template here, that WAS explicitly dependent on the last msg publish, would it still hit this?

**Comment by colinl** [comment]
As I said, when the browser is refreshed or the page with template is re-opened, the last message sent to the template is sent again, but the technique that you have used to pick up the message responds only to messages when they are initially sent, not when they are sent on reload.  Read the section Receiving Data in the template docs. https://dashboard.flowfuse.com/nodes/widgets/ui-template.

**Comment by omrid01** [comment]
> [@omrid01](https://github.com/omrid01) I am trying to understand whether there is anything specific that needs changing here in order to determine whether it will be fixed by the ongoing work in [#2234](https://github.com/FlowFuse/node-red-dashboard/issues/2234).

I like the idea of an 'out of the box' data layer between the node and its widgets, as described in [#2234](https://github.com/FlowFuse/node-red-dashboard/issues/2234). 
This can work well in a _**Shared**_ mode, where the node widgets show the same data across all open client browsers. Data messages coming from the flow are broadcasted by the node to its widgets across all clients, while automatically updating the data layer to allow reconstruction of the widget in newly-opened or refreshed clients. So far so good.

But there are some gaps, as not all messages are born equal:

1. The data snapshot of a widget can be more than just the last received message. It could have been composed by multiple messages. Also, sometimes we wish to save some history, like in a chart. And if we try to save and replay message sequences, then how far back do we go? and what about control and "action" messages, which will be re-executed?
2. This doesn't fully cover client-originated messages. When a button sends an add/change/delete message to a widget, the widget should update the data layer. But some commands are client-specific by nature, e.g. input field values, sort/filter/hide/show in a table etc.

I believe there should be a way to distinguish data-affecting messages from others. I suggest enabling an optional per-message property for the developer to determine if it should update the data layer (and other clients) or not.

Note, that when it comes to **_Multi-user_** mode, where every widget can have its own data image across clients, a data layer concept becomes very complex and impractical, as you need to store a separate data image for each client. In my table node, the data layer is disabled in multi-user mode.

**Comment by colinl** [comment]
@omrid01 can I suggest that you join the discussion there?

**Comment by omrid01** [comment]
> [@omrid01](https://github.com/omrid01) can I suggest that you join the discussion there?

OK. will do.
